File: src/dynamics/pid.c
// SPDX-License-Identifier: Apache-2.0
/*
 * Copyright 2022 Martin Schröder <info@swedishembedded.com>
 * Consulting: https://swedishembedded.com/go
 * Training: https://swedishembedded.com/tag/training
 */

#include "control/dynamics.h"

#include <math.h>
#include <string.h>

void pid_init(struct pid *self)
{
	memset(self, 0, sizeof(*self));
}

void pid_set_from_imc(struct pid *self, float aggr, float process_gain, float time_constant,
		      float dead_time, enum pid_imc_mode mode)
{
	float tc = 0;

	switch (mode) {
	case PID_IMC_AGGRESSIVE:
		tc = fmaxf((0.1F * aggr) * time_constant, (0.8F * aggr) * dead_time);
		break;
	case PID_IMC_MODERATE:
		tc = fmaxf((1.0F * aggr) * time_constant, (8.0F * aggr) * dead_time);
		break;
	case PID_IMC_CONSERVATIVE:
		tc = fmaxf((10.0F * aggr) * time_constant, (80.0F * aggr) * dead_time);
		break;
	}

	float kc = (1.0F / process_gain) *
		   ((time_constant + 0.5F * dead_time) / (tc + 0.5F * dead_time));
	float ti = time_constant + 0.5F * dead_time;
	float td = (time_constant * dead_time) / (2.0F * time_constant + dead_time);

	self->Kp = kc;
	self->Ki = kc / ti;
	self->Kd = kc * td;
}

void pid_set_gains(struct pid *self, float Kp, float Ki, float Kd, float d)
{
	self->Kp = Kp;
	self->Ki = Ki;
	self->Kd = Kd;
	self->d = d;
}

float pid_step(struct pid *self, float e)
{
	float Kp = self->Kp;
	float Ki = self->Ki;
	float Kd = self->Kd;
	float d = self->d;
	float gain = -Kd * d + Kd + Kp;

	self->x[2] = self->x[1];
	self->x[1] = self->x[0];
	self->x[0] = e;
	self->y[2] = self->y[1];
	self->y[1] = self->y[0];
	self->y[0] = 0;
	self->y[0] += self->x[0] * (1) * gain;
	self->y[0] += self->x[1] *
		      ((2 * Kd * d - 2 * Kd + Ki - Kp * d - Kp) / (-Kd * d + Kd + Kp)) * gain;
	self->y[0] += self->x[2] * ((-Kd * d + Kd - Ki * d + Kp * d) / (-Kd * d + Kd + Kp)) * gain;
	self->y[0] -= self->y[1] * (-d - 1);
	self->y[0] -= self->y[2] * (d);
	return self->y[0];
}

File: tests/dynamics/pid.m
% SPDX-License-Identifier: MIT
% Copyright 2022 Martin Schröder <info@swedishembedded.com>
% Consulting: https://swedishembedded.com/go
% Training: https://swedishembedded.com/tag/training

pkg load control
pkg load symbolic

output_precision(9);

Ts = 0.1

function [p, i, d] = pid_imc(aggr, process_gain, time_constant, dead_time)
    tc = max((0.1 * aggr) * time_constant, (0.8 * aggr) * dead_time);
    kc = (1.0/process_gain)*((time_constant + 0.5 * dead_time)/(tc + 0.5 * dead_time));
    ti = time_constant + 0.5 * dead_time;
    td = (time_constant * dead_time)/(2.0 * time_constant + dead_time);

    p = kc;
    i = kc / ti;
    d = kc * td;
end

s = tf('s');
z = tf('z', Ts);

% System dynamics
G = 4/(0.4*s + 1);
Gd = ss(c2d(G, Ts))
Gd.a
Gd.b
Gd.c
Gd.d
% calculate ontinuous time gains
[Kp, Ki, Kd] = pid_imc(0.2, 4, 0.4, 0)
C = Kp + Kd * s + Ki / s

% convert into discrete time
Cd = c2d(C, Ts)
sys_Cd = ss(Cd)
sys_Cd.b
sys_Cd.d
